### DISCUSSION TITLE
bzlmod checksum database / transparency log

### DISCUSSION THREAD

**Discussion by jayconrod** [bazelbuild/bazel#23503] [Ideas]
Coming from Go modules, I'm finding that Bazel modules lack an important security feature: a checksum database, also known as a transparency log.

A checksum database would contain a cryptographic hash of every publicly downloadable artifact in the Bazel module ecosystem. The hashes are collected into an append-only cryptographic log and signed by a key trusted by Bazel. Bazel would use the checksum database to verify the integrity of a newly downloaded file before adding its hash to `MODULE.bazel.lock`.

The checksum database essentially would act as a global `MODULE.bazel.lock` file. New modules are still trusted at time of first use, but the "first use" is when the module is first added to the registry, not the "first use" of every Bazel user. This makes it much more difficult for an attacker to manipulate the content of a module from the perspective of an individual user.

To some extent, a checksum database is needed less urgently by Bazel than it was for Go because Bazel's central registry is *centralized*; in Go, a module author makes a module available to users simply by making their repository public. However, the need is still there: the BCR is based on Git, which doesn't provide the cryptographic guarantees needed for this type of log. If an attacker gained write access to the BCR repository, they could change the content of existing modules in a way that Bazel couldn't easily notice.

Some useful references from Go:

- [Certificate Transparency](https://en.wikipedia.org/wiki/Certificate_Transparency) was the inspiration for Go's checksum database.
- [sum.golang.org](https://sum.golang.org) has some information.
- The [Go Modules Reference](https://go.dev/ref/mod#checksum-database) explains the protocol and the usage.
- [Transparent Logs for Skeptical Clients](https://research.swtch.com/tlog) explains the main data structure.

**Comment by fmeum** [comment]
@meteorcloudy @Wyverald @meisterT

**Comment by meteorcloudy** [comment]
Would this already be covered by https://github.com/bazelbuild/bazel-central-registry/discussions/2721?

**Reply by jayconrod** [reply]
Maybe... That proposal is quite heavy on detail, and I'm not sure I understand the threats it guards against at a higher conceptual level.

As I understand, that proposal would require authors to sign artifacts and include the signatures with releases. The signatures are done with sigstore and included in a public transparency log, so clients could automatically detect an artifact that's been tampered with.

The Go approach doesn't require authors to do anything. Instead, an independent service watches the module registry for updates, retrieves release artifacts, and appends cryptographic sums to a transparency log signed with the service's key (which the client trusts).
